Korean language support?

Hello.
I tried to add a command to plugin.yml like:
```
commands:
  퀘스트:
    description: 퀘스트 명령어 입니다.
```

But when I try to build, It occurs an error: 
```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':generatePluginMeta'.
> java.nio.charset.UnmappableCharacterException: Input length = 2
```

Sorry for my bad English :(

--- a/src/main/groovy/ru/endlesscode/bukkitgradle/meta/MetaFile.groovy
+++ b/src/main/groovy/ru/endlesscode/bukkitgradle/meta/MetaFile.groovy
@@ -12,23 +12,23 @@ import java.nio.file.Path
 class MetaFile {
     public static final String NAME = "plugin.yml"
 
-    private static final String[] ATTRIBUTES = [
+    private static final String[] KNOWN_FIELDS = [
             "name", "description", "version", "author", "authors", "website", "main"
     ]
 
-    final List<String> metaLines
-    final List<String> staticLines
-    final Project project
+    final List<String> metaLines = []
+    final List<String> staticLines = []
 
     private final PluginMeta meta
     private final Path metaFile
 
-    MetaFile(Project project, Path file = null) {
-        this.metaLines = []
-        this.staticLines = []
-        this.project = project
-        this.meta = project.bukkit.meta
-        this.metaFile = file ?: this.findMetaFile()
+    MetaFile(Project project) {
+        this(project.bukkit.meta as PluginMeta, findMetaFile(project))
+    }
+
+    MetaFile(PluginMeta meta, Path file) {
+        this.meta = meta
+        this.metaFile = file
 
         this.filterMetaLines()
     }
@@ -38,7 +38,7 @@ class MetaFile {
      *
      * @return The File
      */
-    private Path findMetaFile() {
+    private static Path findMetaFile(Project project) {
         def javaPlugin = project.convention.getPlugin(JavaPluginConvention)
         def mainSourceSet = javaPlugin.sourceSets.main
         def resourceDir = mainSourceSet.resources.srcDirs[0].toPath()
@@ -59,7 +59,7 @@ class MetaFile {
     }
 
     /**
-     * Validates that meta contains all required attributes
+     * Validates that meta contains all required fields
      * If it isn't throw GradleException
      *
      * @param metaItems List of MetaItem
@@ -110,8 +110,8 @@ class MetaFile {
      * @return true if line is dynamic meta, otherwise false
      */
     private static boolean isMetaLine(String line) {
-        for (String attribute in ATTRIBUTES) {
-            if (line.startsWith("$attribute:")) {
+        for (String field in KNOWN_FIELDS) {
+            if (line.startsWith("$field:")) {
                 return true
             }
         }

--- a/src/main/groovy/ru/endlesscode/bukkitgradle/meta/MetaFile.groovy
+++ b/src/main/groovy/ru/endlesscode/bukkitgradle/meta/MetaFile.groovy
@@ -16,8 +16,8 @@ class MetaFile {
             "name", "description", "version", "author", "authors", "website", "main"
     ]
 
-    final List<String> metaLines = []
-    final List<String> staticLines = []
+    private final List<String> metaLines = []
+    private final List<String> extraLines = []
 
     private final PluginMeta meta
     private final Path metaFile
@@ -29,8 +29,6 @@ class MetaFile {
     MetaFile(PluginMeta meta, Path file) {
         this.meta = meta
         this.metaFile = file
-
-        this.filterMetaLines()
     }
 
     /**
@@ -51,11 +49,11 @@ class MetaFile {
      * @param target
      */
     void writeTo(Path target) {
-        this.validateMeta()
-        this.filterMetaLines()
-        this.generateMetaLines()
+        validateMeta()
+        filterMetaLines()
+        generateMetaLines()
 
-        writeLinesTo(target, metaLines, staticLines)
+        writeLinesTo(target, metaLines, extraLines)
     }
 
     /**
@@ -73,24 +71,24 @@ class MetaFile {
     }
 
     /**
-     * Removes all meta lines from metaFile, and saves static lines to
-     * list. If metaFile not exists only clears staticLines
+     * Removes all meta lines from metaFile, and saves extra lines to
+     * list. If metaFile not exists only clears extra lines.
      */
     private void filterMetaLines() {
-        staticLines.clear()
+        extraLines.clear()
         if (Files.notExists(metaFile)) {
             return
         }
 
         metaFile.eachLine { line ->
-            if (isStaticLine(line)) {
-                staticLines << line
+            if (isExtraLine(line)) {
+                extraLines << line
             }
 
             return
         }
 
-        writeLinesTo(metaFile, staticLines)
+        writeLinesTo(metaFile, extraLines)
     }
 
     /**
@@ -99,8 +97,8 @@ class MetaFile {
      * @param line The line to check
      * @return true if line is static
      */
-    private boolean isStaticLine(String line) {
-        return !isMetaLine(line) && (!line.empty || !staticLines.empty)
+    private boolean isExtraLine(String line) {
+        return !(line.empty && extraLines.empty) && !isMetaLine(line)
     }
 
     /**

--- a/src/main/groovy/ru/endlesscode/bukkitgradle/meta/MetaFile.groovy
+++ b/src/main/groovy/ru/endlesscode/bukkitgradle/meta/MetaFile.groovy
@@ -80,7 +80,7 @@ class MetaFile {
             return
         }
 
-        metaFile.eachLine { line ->
+        metaFile.eachLine("UTF-8") { line ->
             if (isExtraLine(line)) {
                 extraLines << line
             }
@@ -108,13 +108,7 @@ class MetaFile {
      * @return true if line is dynamic meta, otherwise false
      */
     private static boolean isMetaLine(String line) {
-        for (String field in KNOWN_FIELDS) {
-            if (line.startsWith("$field:")) {
-                return true
-            }
-        }
-
-        return false
+        return KNOWN_FIELDS.any { line.startsWith("$it:") }
     }
 
     /**
@@ -137,9 +131,7 @@ class MetaFile {
      */
     private static void writeLinesTo(Path target, List<String>... lineLists) {
         target.withWriter("UTF-8") { writer ->
-            lineLists.each { lineList ->
-                lineList.each { line -> writer.println line }
-            }
+            lineLists.flatten().each { line -> writer.println line }
         }
     }
 }

--- a/src/main/groovy/ru/endlesscode/bukkitgradle/server/ServerProperties.groovy
+++ b/src/main/groovy/ru/endlesscode/bukkitgradle/server/ServerProperties.groovy
@@ -22,7 +22,7 @@ class ServerProperties {
     ServerProperties(Path projectPath) {
         propertiesFile = projectPath.resolve(NAME)
         if (Files.exists(propertiesFile)) {
-            properties.load(propertiesFile.newReader())
+            properties.load(propertiesFile.newReader("UTF-8"))
         } else {
             loadDefaults(projectPath.resolve("build").toAbsolutePath())
         }
@@ -34,7 +34,7 @@ class ServerProperties {
         setDefault(devServerDir, defaultPath.resolve("server").toString())
         setDefault(buildToolsDir, defaultPath.resolve("buildtools").toString())
 
-        properties.store(propertiesFile.newWriter(), $/
+        properties.store(propertiesFile.newWriter("UTF-8"), $/
  This file should *NOT* be checked into Version Control Systems,
  as it contains information specific to your local configuration./$)
     }

--- a/src/main/groovy/ru/endlesscode/bukkitgradle/server/idea/IdeaRunConfigurationBuilder.groovy
+++ b/src/main/groovy/ru/endlesscode/bukkitgradle/server/idea/IdeaRunConfigurationBuilder.groovy
@@ -26,7 +26,7 @@ class IdeaRunConfigurationBuilder {
         def jvmArgs = runConfiguration.buildJvmArgs(false)
 
         def runConfigurationFile = configurationDir.resolve("${taskName.replace(' ', '_')}.xml")
-        def xml = new MarkupBuilder(runConfigurationFile.newWriter())
+        def xml = new MarkupBuilder(runConfigurationFile.newWriter("UTF-8"))
         xml.component(name: 'ProjectRunConfigurationManager') {
             configuration(
                     default: 'false',

--- a/src/main/groovy/ru/endlesscode/bukkitgradle/server/task/PrepareServer.groovy
+++ b/src/main/groovy/ru/endlesscode/bukkitgradle/server/task/PrepareServer.groovy
@@ -38,9 +38,9 @@ class PrepareServer extends DefaultTask {
         }
 
         Properties properties = new Properties()
-        properties.load(eulaFile.newReader())
+        properties.load(eulaFile.newReader("UTF-8"))
         properties.setProperty("eula", "${this.run.eula}")
-        properties.store(eulaFile.newWriter(), "By changing the setting below to TRUE you are indicating your agreement to our EULA (https://account.mojang.com/documents/minecraft_eula).")
+        properties.store(eulaFile.newWriter("UTF-8"), "By changing the setting below to TRUE you are indicating your agreement to our EULA (https://account.mojang.com/documents/minecraft_eula).")
     }
 
     void resolveOnlineMode() {
@@ -50,9 +50,9 @@ class PrepareServer extends DefaultTask {
         }
 
         Properties properties = new Properties()
-        properties.load(propsFile.newReader())
+        properties.load(propsFile.newReader("UTF-8"))
         properties.setProperty("online-mode", "${this.run.onlineMode}")
-        properties.store(propsFile.newWriter(), "Minecraft server properties")
+        properties.store(propsFile.newWriter("UTF-8"), "Minecraft server properties")
     }
 
     void copyPluginsToServerDir() {

--- a/src/test/groovy/ru/endlesscode/bukkitgradle/PluginTestBase.groovy
+++ b/src/test/groovy/ru/endlesscode/bukkitgradle/PluginTestBase.groovy
@@ -50,26 +50,6 @@ class PluginTestBase {
         }
     }
 
-    protected Path createDefaultMetaFile() {
-        Path metaDir = project.buildDir.toPath().resolve("meta/")
-        Files.createDirectories(metaDir)
-        Path metaFile = metaDir.resolve(MetaFile.NAME)
-        Files.deleteIfExists(metaFile)
-        Files.createFile(metaFile)
-
-        metaFile << '''name: TestPlugin
-description: Test plugin description
-version: 0.1
-
-main: com.example.plugin.Plugin
-author: OsipXD
-website: www.example.com
-
-depend: [Vault, ProtocolLib]
-command:
-  example'''
-    }
-
     protected void executeTask(Task task) {
         task.taskDependencies.getDependencies(task).each {
             subTask -> executeTask(subTask)

--- a/src/test/groovy/ru/endlesscode/bukkitgradle/meta/MetaFileTest.groovy
+++ b/src/test/groovy/ru/endlesscode/bukkitgradle/meta/MetaFileTest.groovy
@@ -2,32 +2,49 @@ package ru.endlesscode.bukkitgradle.meta
 
 import org.gradle.api.GradleException
 import org.junit.Before
+import org.junit.Rule
 import org.junit.Test
-import ru.endlesscode.bukkitgradle.PluginTestBase
+import org.junit.rules.TemporaryFolder
+import ru.endlesscode.bukkitgradle.meta.extension.PluginMeta
 
 import java.nio.file.Path
 
-class MetaFileTest extends PluginTestBase {
-    private MetaFile metaFile
+class MetaFileTest {
+
+    @Rule
+    public TemporaryFolder tempFolder = new TemporaryFolder()
+
     private Path target
+    private PluginMeta meta
+
+    // SUT
+    private MetaFile metaFile
 
     @Before
-    void setUp() throws Exception {
-        super.setUp()
-
-        this.target = this.createDefaultMetaFile()
-        this.metaFile = new MetaFile(this.project, this.target)
+    void setUp() {
+        target = createDefaultMetaFile()
+        meta = new PluginMeta()
+        metaFile = new MetaFile(meta, target)
     }
 
     @Test
-    void testRemovingMetaLines() throws Exception {
-        assert ["depend: [Vault, ProtocolLib]", "command:", "  example"] == target.readLines()
+    void 'when initialized - should keep only unsupported lines'() {
+        // When
+        def lines = target.readLines()
+
+        // Then
+        assert ["depend: [Vault, ProtocolLib]", "command:", "  example"] == lines
     }
 
     @Test
-    void testGeneratingMeta() throws Exception {
-        this.initBukkitMeta()
+    void 'when write meta to file - and meta configured - should write all lines'() {
+        // Given
+        configureMeta()
 
+        // When
+        metaFile.writeTo(target)
+
+        // Then
         List<String> expected = [
                 "name: TestPlugin",
                 "description: Test plugin description",
@@ -39,17 +56,48 @@ class MetaFileTest extends PluginTestBase {
                 "command:",
                 "  example"
         ]
-
-        this.metaFile.writeTo(target)
         assert expected == target.readLines()
     }
 
     @Test(expected = GradleException)
-    void testMissingRequiredMetaMustThrowException() throws Exception {
-        project.bukkit {
-            meta.main = null
-        }
+    void 'when write meta to file - and required field missing - should throw exception'() {
+        // Given
+        meta.main = null
 
-        this.metaFile.writeTo(target)
+        // When
+        metaFile.writeTo(target)
+    }
+
+    private Path createDefaultMetaFile() {
+        Path metaFile = tempFolder.newFile(MetaFile.NAME).toPath()
+        metaFile << $/
+            name: TestPlugin
+            description: Test plugin description
+            version: 0.1
+            
+            main: com.example.plugin.Plugin
+            author: OsipXD
+            website: www.example.com
+            
+            depend: [Vault, ProtocolLib]
+            command:
+              example
+            /$.stripIndent()
+    }
+
+    private void configureMeta(
+            String name = "TestPlugin",
+            String description = "Test plugin description",
+            String version = "0.1",
+            String main = "com.example.plugin.Plugin",
+            List<String> authors = ["OsipXD", "Contributors"],
+            String url = "http://www.example.com/"
+    ) {
+        meta.name = name
+        meta.description = description
+        meta.version = version
+        meta.main = main
+        meta.authors = authors
+        meta.url = url
     }
 }

--- a/src/test/groovy/ru/endlesscode/bukkitgradle/meta/MetaFileTest.groovy
+++ b/src/test/groovy/ru/endlesscode/bukkitgradle/meta/MetaFileTest.groovy
@@ -1,12 +1,16 @@
 package ru.endlesscode.bukkitgradle.meta
 
+import groovy.transform.NamedVariant
 import org.gradle.api.GradleException
 import org.junit.Before
 import org.junit.Rule
 import org.junit.Test
 import org.junit.rules.TemporaryFolder
 import ru.endlesscode.bukkitgradle.meta.extension.PluginMeta
+import ru.endlesscode.bukkitgradle.util.CharsetUtils
 
+import java.lang.reflect.Field
+import java.nio.charset.Charset
 import java.nio.file.Path
 
 class MetaFileTest {
@@ -27,12 +31,13 @@ class MetaFileTest {
         target = tempFolder.newFile(MetaFile.NAME).toPath()
         meta = new PluginMeta()
         metaFile = new MetaFile(meta, source)
+
+        configureMeta()
     }
 
     @Test
     void 'when write meta to file - should keep only unsupported lines in source file'() {
         // Given
-        configureMeta()
         source << $/
             name: TestPlugin
             description: Test plugin description
@@ -55,9 +60,13 @@ class MetaFileTest {
     }
 
     @Test
-    void 'when write meta to file - and meta configured - should write all lines'() {
+    void 'when write meta to file - and all meta configured - should write all lines'() {
         // Given
-        configureMeta()
+        configureMeta(
+                description: "Test plugin description",
+                authors: ["OsipXD", "Contributors"],
+                url: "http://www.example.com/"
+        )
 
         // When
         metaFile.writeTo(target)
@@ -77,7 +86,6 @@ class MetaFileTest {
     @Test
     void 'when write meta to file - and there are extra fields in source - should write all lines'() {
         // Given
-        configureMeta()
         source << $/
             depend: [Vault, ProtocolLib]
             command:
@@ -90,14 +98,37 @@ class MetaFileTest {
         // Then
         List<String> expected = [
                 "name: TestPlugin",
-                "description: Test plugin description",
                 "main: com.example.plugin.Plugin",
                 "version: 0.1",
-                "website: http://www.example.com/",
-                "authors: [OsipXD, Contributors]",
                 "depend: [Vault, ProtocolLib]",
                 "command:",
                 "  example"
+        ]
+        assert expected == target.readLines()
+    }
+
+    @Test
+    void 'when write meta to file - and there are exotic chars in source - should read it correctly'() {
+        // Given
+        source.append($/
+            commands:
+              퀘스트:
+                description: 퀘스트 명령어 입니다.
+            /$.stripIndent())
+
+        // When
+        CharsetUtils.setDefaultCharset('CP866')
+        metaFile.writeTo(target)
+        CharsetUtils.setDefaultCharset('UTF-8')
+
+        // Then
+        List<String> expected = [
+                "name: TestPlugin",
+                "main: com.example.plugin.Plugin",
+                "version: 0.1",
+                "commands:",
+                "  퀘스트:",
+                "    description: 퀘스트 명령어 입니다."
         ]
         assert expected == target.readLines()
     }
@@ -111,18 +142,19 @@ class MetaFileTest {
         metaFile.writeTo(target)
     }
 
+    @NamedVariant
     private void configureMeta(
-            name = "TestPlugin",
-            description = "Test plugin description",
-            version = "0.1",
-            main = "com.example.plugin.Plugin",
-            authors = ["OsipXD", "Contributors"],
-            url = "http://www.example.com/"
+            name = null,
+            description = null,
+            version = null,
+            main = null,
+            authors = null,
+            url = null
     ) {
-        meta.name = name
+        meta.name = name ?: "TestPlugin"
         meta.description = description
-        meta.version = version
-        meta.main = main
+        meta.version = version ?: "0.1"
+        meta.main = main ?: "com.example.plugin.Plugin"
         meta.authors = authors
         meta.url = url
     }

--- a/src/test/groovy/ru/endlesscode/bukkitgradle/meta/MetaFileTest.groovy
+++ b/src/test/groovy/ru/endlesscode/bukkitgradle/meta/MetaFileTest.groovy
@@ -14,6 +14,7 @@ class MetaFileTest {
     @Rule
     public TemporaryFolder tempFolder = new TemporaryFolder()
 
+    private Path source
     private Path target
     private PluginMeta meta
 
@@ -22,24 +23,66 @@ class MetaFileTest {
 
     @Before
     void setUp() {
-        target = createDefaultMetaFile()
+        source = tempFolder.newFile("source-$MetaFile.NAME").toPath()
+        target = tempFolder.newFile(MetaFile.NAME).toPath()
         meta = new PluginMeta()
-        metaFile = new MetaFile(meta, target)
+        metaFile = new MetaFile(meta, source)
     }
 
     @Test
-    void 'when initialized - should keep only unsupported lines'() {
+    void 'when write meta to file - should keep only unsupported lines in source file'() {
+        // Given
+        configureMeta()
+        source << $/
+            name: TestPlugin
+            description: Test plugin description
+            version: 0.1
+            
+            main: com.example.plugin.Plugin
+            author: OsipXD
+            website: www.example.com
+            
+            depend: [Vault, ProtocolLib]
+            command:
+              example
+            /$.stripIndent()
+
         // When
-        def lines = target.readLines()
+        metaFile.writeTo(target)
 
         // Then
-        assert ["depend: [Vault, ProtocolLib]", "command:", "  example"] == lines
+        assert ["depend: [Vault, ProtocolLib]", "command:", "  example"] == source.readLines()
     }
 
     @Test
     void 'when write meta to file - and meta configured - should write all lines'() {
         // Given
         configureMeta()
+
+        // When
+        metaFile.writeTo(target)
+
+        // Then
+        List<String> expected = [
+                "name: TestPlugin",
+                "description: Test plugin description",
+                "main: com.example.plugin.Plugin",
+                "version: 0.1",
+                "website: http://www.example.com/",
+                "authors: [OsipXD, Contributors]"
+        ]
+        assert expected == target.readLines()
+    }
+
+    @Test
+    void 'when write meta to file - and there are extra fields in source - should write all lines'() {
+        // Given
+        configureMeta()
+        source << $/
+            depend: [Vault, ProtocolLib]
+            command:
+              example
+            /$.stripIndent()
 
         // When
         metaFile.writeTo(target)
@@ -68,30 +111,13 @@ class MetaFileTest {
         metaFile.writeTo(target)
     }
 
-    private Path createDefaultMetaFile() {
-        Path metaFile = tempFolder.newFile(MetaFile.NAME).toPath()
-        metaFile << $/
-            name: TestPlugin
-            description: Test plugin description
-            version: 0.1
-            
-            main: com.example.plugin.Plugin
-            author: OsipXD
-            website: www.example.com
-            
-            depend: [Vault, ProtocolLib]
-            command:
-              example
-            /$.stripIndent()
-    }
-
     private void configureMeta(
-            String name = "TestPlugin",
-            String description = "Test plugin description",
-            String version = "0.1",
-            String main = "com.example.plugin.Plugin",
-            List<String> authors = ["OsipXD", "Contributors"],
-            String url = "http://www.example.com/"
+            name = "TestPlugin",
+            description = "Test plugin description",
+            version = "0.1",
+            main = "com.example.plugin.Plugin",
+            authors = ["OsipXD", "Contributors"],
+            url = "http://www.example.com/"
     ) {
         meta.name = name
         meta.description = description

--- a/src/test/groovy/ru/endlesscode/bukkitgradle/util/CharsetUtil.groovy
+++ b/src/test/groovy/ru/endlesscode/bukkitgradle/util/CharsetUtil.groovy
@@ -1,0 +1,14 @@
+package ru.endlesscode.bukkitgradle.util
+
+import java.lang.reflect.Field
+import java.nio.charset.Charset
+
+class CharsetUtils {
+
+    static void setDefaultCharset(String charset) {
+        System.setProperty("file.encoding", charset)
+        Field charsetField = Charset.getDeclaredField("defaultCharset")
+        charsetField.setAccessible(true)
+        charsetField.set(null, null)
+    }
+}

